substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/hosted/JNINativeCallWrapperSubstitutionProcessor.java

```diff
@@ -41,6 +42,13 @@
     @Override
     public ResolvedJavaMethod lookup(ResolvedJavaMethod method) {
         assert method.isNative() : "Must have been registered as a native substitution processor";
+        if (method.getDeclaringClass().getName().equals("Lcom/oracle/svm/jni/JNIJavaCallWrappers;")) {
+            // Avoid generating JNINativeCallWrapperMethods for trampolines
+            JNICallTrampolineMethod callTrampolineMethod = JNIAccessFeature.singleton().getCallTrampolineMethod(method.getName());
+            if (callTrampolineMethod != null) {
+                return callTrampolineMethod;
+            }
```

Comment:
Good point. As you have [previously suggested](https://github.com/oracle/graal/issues/4018#issuecomment-991098259) we can determine whether we should generate a `JNINativeCallWrapperMethod` or not even if the trampolines have not been created yet by having the trampolines enumerated or defined in a different class. However, what should `JNINativeCallWrapperSubstitutionProcessor#lookup` return in that case? Would it be OK to return `null` (making the necessary adjustments in the callers to handle it)?

Reply:
ping @peter-hofer

Reply:
The `JNICallTrampolineMethod` should always be created in `JNINativeCallWrapperSubstitutionProcessor.lookup`. The method that supersedes `JNIAccessFeature.createJavaCallTrampoline` and registers the trampoline for compilation can just use the `MetaAccessProvider` as follows, which will eventually delegate to `JNINativeCallWrapperSubstitutionProcessor`:
```
access.getMetaAccess().lookupJavaMethod(ReflectionUtil.lookupMethod(JNIJavaCallTrampolines.class, "arrayJavaCallTrampoline"))
```

Reply:
> The method that supersedes `JNIAccessFeature.createJavaCallTrampoline` and registers the trampoline for compilation...

AFAIU `JNIAccessFeature.createJavaCallTrampoline` is the one registering the trampoline for compilation.

> ... can just use the `MetaAccessProvider` as follows, which will eventually delegate to `JNINativeCallWrapperSubstitutionProcessor`

Unfortunately this doesn't seem to be true. The method is getting resolved without going through the substitution processor:

```
fromMetaspace:115, HotSpotResolvedJavaMethodImpl (jdk.vm.ci.hotspot)
asResolvedJavaMethod:-1, CompilerToVM (jdk.vm.ci.hotspot)
lookupJavaMethod:81, HotSpotMetaAccessProvider (jdk.vm.ci.hotspot)
lookupJavaMethod:63, HotSpotSnippetMetaAccessProvider (org.graalvm.compiler.hotspot)
createJavaCallTrampoline:194, JNIAccessFeature (com.oracle.svm.jni.access)
beforeAnalysis:171, JNIAccessFeature (com.oracle.svm.jni.access)
lambda$runPointsToAnalysis$9:694, NativeImageGenerator (com.oracle.svm.hosted)
accept:-1, 194672584 (com.oracle.svm.hosted.NativeImageGenerator$$Lambda$399)
forEachFeature:74, FeatureHandler (com.oracle.svm.hosted)
runPointsToAnalysis:694, NativeImageGenerator (com.oracle.svm.hosted)
doRun:536, NativeImageGenerator (com.oracle.svm.hosted)
run:494, NativeImageGenerator (com.oracle.svm.hosted)
buildImage:426, NativeImageGeneratorRunner (com.oracle.svm.hosted)
build:587, NativeImageGeneratorRunner (com.oracle.svm.hosted)
main:126, NativeImageGeneratorRunner (com.oracle.svm.hosted)
main:617, NativeImageGeneratorRunner$JDK9Plus (com.oracle.svm.hosted)
```

Reply:
> AFAIU `JNIAccessFeature.createJavaCallTrampoline` is the one registering the trampoline for compilation.

Exactly. Looking up a method (creating some instance of `ResolvedJavaMethod`) and making it reachable (in this case manually registering it for compilation) are different concerns. `JNINativeCallWrapperSubstitutionProcessor` can create the `JNICallTrampolineMethod` method when it is looked up, but it still won't be reachable, so `JNIAccessFeature` needs to still register it explicitly.

> Unfortunately this doesn't seem to be true. The method is getting resolved without going through the substitution processor:
> `lookupJavaMethod:63, HotSpotSnippetMetaAccessProvider (org.graalvm.compiler.hotspot)`
> `createJavaCallTrampoline:194, JNIAccessFeature (com.oracle.svm.jni.access)`

You need to use `BeforeAnalysisAccessImpl.getMetaAccess()` which returns an `AnalysisMetaAccess`. It looks like using the wrapped provider was a workaround to specifically avoid `JNINativeCallWrapperSubstitutionProcessor` and get a JVMCI `ResolvedJavaMethod` instead of a `JNINativeCallWrapperMethod`, i.e. the very thing you are fixing/cleaning up with this change.
I'd wager you can also use the `AnalysisMetaAccess` for the field:
```
access.registerAsAccessed((AnalysisField) JNIAccessibleMethod.getCallWrapperField(access.getMetaAccess(), variant, nonVirtual));
```

Reply:
> It looks like using the wrapped provider was a workaround to specifically avoid JNINativeCallWrapperSubstitutionProcessor and get a JVMCI ResolvedJavaMethod instead of a JNINativeCallWrapperMethod, i.e. the very thing you are fixing/cleaning up with this change.

That was usefull :)

> I'd wager you can also use the AnalysisMetaAccess for the field:

I am getting the following error when using the `AnalysisMetaAccess` for the field so I kept the wrapped one:

```
Fatal error: java.lang.ClassCastException: class jdk.vm.ci.hotspot.HotSpotResolvedJavaMethodImpl cannot be cast to class com.oracle.graal.pointsto.meta.AnalysisMethod (jdk.vm.ci.hotspot.HotSpotResolvedJavaMethodImpl is in module jdk.internal.vm.ci of loader 'bootstrap'; com.oracle.graal.pointsto.meta.AnalysisMethod is in unnamed module of loader 'app')
	at com.oracle.graal.pointsto.meta.AnalysisType.resolveConcreteMethod(AnalysisType.java:844)
	at com.oracle.graal.pointsto.meta.AnalysisType.resolveConcreteMethod(AnalysisType.java:73)
	at com.oracle.graal.pointsto.meta.AnalysisType.resolveConcreteMethod(AnalysisType.java:851)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:622)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:612)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:612)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:612)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:612)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:612)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.getMethodImplementations(AnalysisUniverse.java:596)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.collectMethodImplementations(AnalysisUniverse.java:583)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.setAnalysisDataValid(AnalysisUniverse.java:167)
	at com.oracle.graal.pointsto.PointsToAnalysis.finish(PointsToAnalysis.java:651)
	at com.oracle.graal.pointsto.PointsToAnalysis.runAnalysis(PointsToAnalysis.java:734)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:701)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:536)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:494)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:426)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:587)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:126)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:617)
```

Please see the latest commit for those changes and let me know if it looks OK. If things look good I will then squash the fixup commits.